File: src/main/java/me/miki/shindo/utils/OptifineUtils.java
package me.miki.shindo.utils;

import me.miki.shindo.injection.mixin.ShindoTweaker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.settings.GameSettings;

import java.lang.reflect.Field;

public class OptifineUtils {

    private static final Minecraft mc = Minecraft.getMinecraft();
    private static Field gameSettings_ofFastRender;

    static {
        try {
            Class.forName("Config");

            gameSettings_ofFastRender = GameSettings.class.getDeclaredField("ofFastRender");
            gameSettings_ofFastRender.setAccessible(true);
        } catch (ClassNotFoundException | NoSuchFieldException ignore) {
        }
    }

    public static void disableFastRender() {

        if (ShindoTweaker.hasOptifine) {
            try {
                OptifineUtils.gameSettings_ofFastRender.set(mc.gameSettings, false);
            } catch (IllegalArgumentException | IllegalAccessException ignored) {
            }
        }

        mc.gameSettings.useVbo = true;
        mc.gameSettings.fboEnable = true;
    }
}

File: src/main/kotlin/com/shindoclient/shindo/utils/OptifineUtils.kt
package com.shindoclient.shindo.utils

import com.shindoclient.shindo.injection.mixin.ShindoTweaker
import net.minecraft.client.Minecraft
import net.minecraft.client.settings.GameSettings
import java.lang.reflect.Field

object OptifineUtils {
    private val mc: Minecraft = Minecraft.getMinecraft()

    private val gameSettingsOfFastRender: Field? =
        run {
            try {
                Class.forName("Config")
                GameSettings::class.java.getDeclaredField("ofFastRender").apply { isAccessible = true }
            } catch (_: ClassNotFoundException) {
                null
            } catch (_: NoSuchFieldException) {
                null
            }
        }

    @JvmStatic
    fun disableFastRender() {
        if (ShindoTweaker.hasOptifine) {
            try {
                gameSettingsOfFastRender?.set(mc.gameSettings, false)
            } catch (_: IllegalArgumentException) {
            } catch (_: IllegalAccessException) {
            }
        }

        mc.gameSettings.useVbo = true
        mc.gameSettings.fboEnable = true
    }
}
